### DISCUSSION TITLE
Reload configuration without restart?

### DISCUSSION THREAD

**Discussion by donnyxray** [XTLS/Xray-core#1060] [Q&A]
When I add or remove a client/user, I restart xray-core. The problem with this approach is that it breaks ongoing connections for other users. They obviously don't like this, especially when downloading a large file. How do you deal with this?

It seems the api command can add/remove entire inbounds, but not single clients/users?

Ideally xray-core can work like haproxy: when you reload configuration, the old haproxy process lives until all existing connections are closed. Meanwhile new connections are already handled by a new haproxy process. This way the "reload" is just a socket transfer trick and no complicated internal reload handling is needed.

**Comment by ghost** [comment]
Yet, It's a fantastic feature for a powerful proxy application.

**Comment by TheLordOfTheKings** [comment]
@RPRX Can you please make a workaround for not restarting the whole service?

**Comment by shakibamoshiri** [comment]
This is not easy to implement, even HA is under criticism for not doing a hot-reload really well.  
Take this scenario that xray server is holding 1000 websocket connections and an admin reloads it -- what should happen?

xray server should still hold those 1K ws and accepting new incoming ws connections ? sounds simple but over time that xray server has to hold the oldest connection meanwhile taking care of any new connections as well.  
IT takes more CPU and RAM , etc. 

a simpler solution is to dump back the whole config.json file via CLI. 

So when a new inbound is added , an admin by dumping the whole new config, can copy it to **config.json** and reload manually.

**Reply by scriptvpskita** [reply]
how ?

**Reply by shakibamoshiri** [reply]
you can add new inbound via `api` without reload 
what if for new users ? well you add new users to a temporary inbound , then swap the temporary inbound with active inbound (delete old, add new) the new user will be active and no reload/restart is needed

**Reply by scriptvpskita** [reply]
do you have example command or document to do that ?

**Reply by shakibamoshiri** [reply]
you can read the Xray doc and try some tests , I do not have it at the moment

**Reply by scriptvpskita** [reply]
please send a link document

**Reply by shakibamoshiri** [reply]
https://xtls.github.io/config/inbound.html
